### PR TITLE
chore(atomic, headless): make bueno external

### PR DESCRIPTION
**Pull request by fpbrault**
This PR makes bueno external in atomic and headless.

https://coveord.atlassian.net/browse/KIT-3551

### CODE REVIEW

File: packages/atomic/.storybook/main.mts [hunk 1]
```diff
@@ -34,77 +34,21 @@ function externalizeDependencies() {
     name: 'externalize-dependencies',
     enforce: 'pre',
     resolveId: (id: string) => {
-      if (id.startsWith('/headless')) {
+      if (/^\/(headless|bueno)/.test(id)) {
         return false;
       }
-      if (packageMappings[id]) {
+      if (generateExternalPackageMappings(__dirname)[id]) {
         if (!isCDN) {
```
Comment (fbeaudoincoveo) [inline packages/atomic/.storybook/main.mts:41]: Isn't this superfluous? We're already checking `isCDN` here: https://github.com/coveo/ui-kit/pull/4433/files#diff-5dda87fd1db013e00675717666c12d3ee6a53108c325e257ead47ef5b9573d0cR27 and this seems to be the only place where we're calling `generateExternalPackageMappings`, so this condition should never be satisfied.

File: packages/atomic/package.json [hunk 2]
```diff
@@ -141,7 +141,8 @@
     "wait-on": "7.2.0"
   },
   "peerDependencies": {
-    "@coveo/headless": "3.0.0"
+    "@coveo/headless": "2.80.0",
```
Comment (fbeaudoincoveo) [inline packages/atomic/package.json:144]: Why the downgrade?

File: packages/atomic/stencil.config.ts [hunk 3]
```diff
@@ -16,69 +15,19 @@ import html from 'rollup-plugin-html';
...
-    cdn: `/bueno/${buenoVersion}/bueno.esm.js`,
-  }, */
-};
-
 function generateAliasEntries() {
-  return Object.entries(packageMappings).map(([find, paths]) => ({
-    find,
-    replacement: paths.devWatch,
-  }));
+  return Object.entries(generateExternalPackageMappings(__dirname)).map(
```
Comment (fbeaudoincoveo) [inline packages/atomic/stencil.config.ts:25]: In this file, I suggest assigning the result of `generateExternalPackageMappings(__dirname)` to a variable and referencing it here and in lines 217 and 223

**Review by alexprudhomme — COMMENTED** [review]
why is this needed again ? I don't see the point of externalizing bueno. If I were to make a library that uses `zod` for schema validation, I feel like I would bundle those validation functions with my library.

**Comment by fpbrault** [comment]
> why is this needed again ? I don't see the point of externalizing bueno. If I were to make a library that uses `zod` for schema validation, I feel like I would bundle those validation functions with my library. This is not a `framework` headless and atomic depend on to work. This is a library for utility functions.

The main reason to externalize Bueno was to avoid code duplication when a end-user bundles Atomic or Headless in their own project: https://coveord.atlassian.net/browse/KIT-3181 I dont know if there was other reasons @louis-bompart ?

**Comment by alexprudhomme** [comment]
> > why is this needed again ? I don't see the point of externalizing bueno. If I were to make a library that uses `zod` for schema validation, I feel like I would bundle those validation functions with my library. This is not a `framework` headless and atomic depend on to work. This is a library for utility functions.
> 
> The main reason to externalize Bueno was to avoid code duplication when a end-user bundles Atomic or Headless in their own project: https://coveord.atlassian.net/browse/KIT-3181 I dont know if there was other reasons @louis-bompart ?

But do users really install bueno ? If they don't, they should not have duplication right? I feel like bueno is more an internal package of us that we made before there were popular schema validation libraries like `zod`.

**Comment by fpbrault** [comment]
> > > why is this needed again ? I don't see the point of externalizing bueno. If I were to make a library that uses `zod` for schema validation, I feel like I would bundle those validation functions with my library. This is not a `framework` headless and atomic depend on to work. This is a library for utility functions.
> > 
> > 
> > The main reason to externalize Bueno was to avoid code duplication when a end-user bundles Atomic or Headless in their own project: https://coveord.atlassian.net/browse/KIT-3181 I dont know if there was other reasons @louis-bompart ?
> 
> But do users really install bueno ? If they don't, they should not have duplication right? I feel like bueno is more an internal package for us that we made before there were popular schema validation libraries like `zod`.

Thats a good point, I see ~20k weekly downloads on npmjs.org, but I would not be surprised if these are all from us/our CI...

**Comment by louis-bompart** [comment]
> > > > why is this needed again ? I don't see the point of externalizing bueno. If I were to make a library that uses `zod` for schema validation, I feel like I would bundle those validation functions with my library. This is not a `framework` headless and atomic depend on to work. This is a library for utility functions.
> > > 
> > > 
> > > The main reason to externalize Bueno was to avoid code duplication when a end-user bundles Atomic or Headless in their own project: https://coveord.atlassian.net/browse/KIT-3181 I dont know if there was other reasons @louis-bompart ?
> > 
> > 
> > But do users really install bueno ? If they don't, they should not have duplication right? I feel like bueno is more an internal package for us that we made before there were popular schema validation libraries like `zod`.
> 
> Thats a good point, I see ~20k weekly downloads on npmjs.org, but I would not be surprised if these are all from us/our CI...

It is very unlikely that we have that many users for Bueno (, one of these days, we might want to consider putting the 🪓 into Bueno in favour of another mainstream library like Zod.

However, the concept/philosophy here still applies:
 - We want to externalize our dependencies for CDN users to facilitate chunking wherever possible and allow loading dependencies twice. In the present case, we see that Bueno is used by both Headless & Atomic, which means that if we *don't* externalize Bueno, CDN users will load it twice (is that a big deal, not much 'cause Bueno is very, very, very light)
 - We want to externalize our dependencies for NPM users to avoid them bundling the dependencies twice or more. Granted, in the case of Bueno, the chances are minimal, and the gains are abysmal.

File: .github/workflows/prbot.yml [hunk 4]
```diff
@@ -91,7 +91,6 @@ jobs:
       - uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4
       - uses: ./.github/actions/setup
       - run: npm run build
-      - run: cat packages/atomic/dist/atomic/atomic.esm.js
```
Comment (louis-bompart) [inline .github/workflows/prbot.yml:94]: Clean up

File: .github/workflows/prbot.yml [hunk 5]
```diff
@@ -107,7 +106,6 @@ jobs:
...
```
Comment (louis-bompart) [inline .github/workflows/prbot.yml:110]: Clean up

File: packages/atomic/.storybook/main.mts [hunk 6]
```diff
@@ -47,9 +47,12 @@ const config: StorybookConfig = {
...
   },
 
-  viteFinal: async (config) =>
+  viteFinal: async (config, {configType}) =>
     mergeConfig(config, {
-      plugins: [nxViteTsPaths(), externalizeDependencies()],
+      plugins: [
+        nxViteTsPaths(),
+        configType === 'PRODUCTION' && isCDN && externalizeDependencies(),
+      ],
```
Comment (louis-bompart) [inline packages/atomic/.storybook/main.mts:55]: `configType === 'PRODUCTION'` means we run `storybook build`.
We are building the storybook outside the CDN when we run the tests to go fast and leverage build caches.
In this specific case, we do want rollup to be smart and resolve our dependencies (and yes bundle them, but we don't care: it ain't exported)

**Comment by louis-bompart** [comment]
I took the liberty of fixing those poor tests (& removing a few cats from the code 😿). I peppered my changes with comments to explain the gist of it, can talk you thru tmrw if you want!
https://github.com/coveo/ui-kit/pull/4433/commits/d2aa4e5eda30e76d6e385f60f43fb6f2e0d48ce7

**Review by fbeaudoincoveo — APPROVED** [review]

**Pull request merged**